Fix NullReferenceException in DependentProjectsFinder when we have an…

pharring opened the pull request:
… InternalsVisibleTo attribute with a null constructor arg.

Fixes Watson bug 1174256

pharring commented:
@dpoeschl @rchande @jasonmalinowski for review (ask mode)

dpoeschl commented on src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs:
```diff
@@ -364,6 +364,11 @@ private static Lazy<HashSet<string>> CreateInternalsVisibleToMap(IAssemblySymbol
+                    }
+
                     var commaIndex = value.IndexOf(',');
                     var assemblyName = commaIndex >= 0 ? value.Substring(0, commaIndex).Trim() : value;
```
Unrelated, but I wonder if ```value``` should be trimmed as well.

pharring replied:
Yes, I also thought of adding further validation here to check for obviously invalid assembly names (empty, all whitespace, illegal chars, etc.). However, in the spirit of fixing the Watson bug, I left it alone. On that note, one could argue that "IsNullOrWhitespace" is more than required to fix the NRE - a simple null check would be sufficient.

Pilchie replied:
I understand the desire to do the minimum for the watson bug, but did we examine the customer scenario when empty strings are passed, etc?  I wouldn't want to fix this, and then have it crash again two lines further down if it was an empty or whitespace string.

pharring replied:
I did some manual testing around this in EditorTestApp and the only case that causes the NRE is null. Empty string, strings with comma, missing argument, all "work" as expected (it's a compiler error, but the IDE doesn't choke).

Pilchie replied:
Thanks!

dpoeschl commented on src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs:
```diff
@@ -364,6 +364,11 @@ private static Lazy<HashSet<string>> CreateInternalsVisibleToMap(IAssemblySymbol
                     }
 
                     var value = (string)typeNameConstant.Value;
+                    if (string.IsNullOrWhiteSpace(value))
```
The nittiest of nits: Is there any way to have meaningful whitespace here? If the string is ```"<whitespace>,<other stuff>"``` then we'd previously let that ```<whitespace>``` get into the map.

dpoeschl replied:
Er, misspoke. If the string *didn't* contain the comma, then we left the whitespace in. This can't matter, right?!

pharring replied:
If we're strictly fixing the NRE then, yes, you're right. Even a leading comma would have previously resulted in String.Empty going in the map. As I mentioned in reply to your other comment, I was thinking of adding more validation here, but didn't. In that vein,  I'll change this check to a simple null check.

dpoeschl commented on src/EditorFeatures/Test2/FindReferences/FindReferencesTests.NamedTypeSymbols.vb:
```diff
@@ -2156,5 +2156,35 @@ public class D { }
+        <Document><![CDATA[
+        public class B
+        {
+            [$$A] _a;
```
Did you mean ```[|$$A|]``` to indicate that the ```A``` should be found? As it is I think this will result in a C# file with literally ```[A]``` in it.

pharring replied:
Possibly. I tried a few combinations. It doesn't actually matter if the 2nd project compiles, just that it tries to reference something unseen that will exercise the DependentProjectsFinder code (I verified by putting a BP on the new null check). You might have to help me write a suitable test.

pharring commented:
@dpoeschl Thanks for the initial review. I changed the "IsNullOrWhitespace" check to a simple null check. That's the minimum required to fix the bug. I also adjusted the unit test (you were right about the missing vertical bars).

jasonmalinowski commented:
:+1:

rchande commented:
:+1:

pharring commented:
Needs to be ported to Stabilization branch. Will open a separate PR for that.

Pilchie commented:
:+1:

dpoeschl commented:
Late to the party, but :+1: with the fixed unit test.
